hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/KeyDeletingService.java

```diff
@@ -235,27 +236,108 @@ private Pair<Pair<Integer, Long>, Boolean> submitPurgeKeysRequest(List<DeleteBlo
+                .setUuid(HddsUtils.toProtobuf(expectedPreviousSnapshotId))
+                .build());
+      }
+
+      currSize = requestBuilder.build().getSerializedSize();
+
+      // 3.1 Add purgeKeys
+      List<String> batchPurgeKeys = new ArrayList<>();
+
+      while (purgeKeyIndex < purgeKeys.size()) {
+        String nextKey = purgeKeys.get(purgeKeyIndex);
+
+        int estimatedKeySize = estimateStringEntrySize(nextKey);
+
+        if (currSize + estimatedKeySize > ratisLimit) {
+          batchCapacityReached = true;
+          break;
+        }
+
+        currSize += estimatedKeySize;
+        batchPurgeKeys.add(nextKey);
+        purgeKeyIndex++;
       }
-    }
 
-    OzoneManagerProtocolProtos.OMRequest omRequest = OzoneManagerProtocolProtos.OMRequest.newBuilder()
-        .setCmdType(OzoneManagerProtocolProtos.Type.PurgeKeys)
-        .setPurgeKeysRequest(purgeKeysRequest)
-        .setClientId(getClientId().toString())
-        .build();
-
-    // Submit PurgeKeys request to OM. Acquire bootstrap lock when processing deletes for snapshots.
-    try (BootstrapStateHandler.Lock lock = snapTableKey != null ? getBootstrapStateLock().lock() : null) {
-      OzoneManagerProtocolProtos.OMResponse omResponse = submitRequest(omRequest);
-      if (omResponse != null) {
-        purgeSuccess = purgeSuccess && omResponse.getSuccess();
+      // Now actually add batchPurgeKeys
+      if (!batchPurgeKeys.isEmpty()) {
+        OzoneManagerProtocolProtos.DeletedKeys deletedKeys = OzoneManagerProtocolProtos.DeletedKeys.newBuilder()
+            .setVolumeName("").setBucketName("")
+            .addAllKeys(batchPurgeKeys).build();
+        requestBuilder.addDeletedKeys(deletedKeys);
+        currSize = requestBuilder.build().getSerializedSize();
+      }
+
+      // 3.2 Add keysToUpdate
+      while (!batchCapacityReached && updateIndex < keysToUpdateList.size()) {
+        OzoneManagerProtocolProtos.SnapshotMoveKeyInfos nextUpdate = keysToUpdateList.get(updateIndex);
+
+        int estimatedSize = nextUpdate.getSerializedSize();
+
+        if (currSize + estimatedSize > ratisLimit) {
+          batchCapacityReached = true;
+          break;
+        }
+
+        requestBuilder.addKeysToUpdate(nextUpdate);
+        currSize += estimatedSize;
+        updateIndex++;
+      }
+
+      // 3.3 Add renamed keys
+      while (!batchCapacityReached && renameEntriesToBeDeleted != null &&
+          renameIndex < renameEntriesToBeDeleted.size()) {
+        String nextRename = renameEntriesToBeDeleted.get(renameIndex);
+
+        int estimatedSize = estimateStringEntrySize(nextRename);
+
+        if (currSize + estimatedSize > ratisLimit) {
+          break;
+        }
+
+        requestBuilder.addRenamedKeys(nextRename);
+        currSize += estimatedSize;
+        renameIndex++;
+      }
+
+      // Finalize and send this batch
+      OzoneManagerProtocolProtos.OMRequest omRequest = OzoneManagerProtocolProtos.OMRequest.newBuilder()
```

Comment:
We can maybe move the submit part into another function so that we can call it just by passing the requestBuilder